New UseHttps overload with per connection callback

Tratcher opened the pull request:
Fixes #33452, contributes to #33264 (delayed client certs).

There is no new functionality here, this takes an internal API and makes it public (and more user friendly). Customers wanted this in order to A) get access to the ConnectionContext for information like the IPs, and B) to configure delayed client cert negotiation per connection.

~~@dotnet/aspnet-api-review having to pass the callback to the TlsHandshakeCallbackOptions constructor is inconvenient usage compared to assigning it as a property. This pattern was recommended because the callback is required. I recommend we instead validate the property is set inside the UseHttps method that takes the TlsHandshakeCallbackOptions.~~ I removed the constructor.

~~Open question: Should the ServerOptionsSelectionCallback still default to AllowDelayedClientCertificateNegotation = true (new in preview6)? Or should people that want delayed client certs move to the new API?~~ Removed. The new API will be required to opt into delayed client certs.

Tratcher committed:
New UseHttps overload with per connection callback #33452

Tratcher commented:
*Ping*

BrennanConroy commented on src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:
```diff
@@ -259,34 +256,37 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, ServerOpt
-                var tlsOptions = await serverOptionsSelectionCallback(stream, clientHelloInfo, state, cancellationToken);
-                return new (tlsOptions, ClientCertificateMode.DelayCertificate);
-            }
+                context.AllowDelayedClientCertificateNegotation = true; // TODO: do we still need this enabled by default after adding the new overload?
```
Don't we want to discourage the usage of delayed cert negotiation? So we should have it off by default IMO.

Tratcher replied:
If we turn it off here then it won't be usable at all from this UseHttps overload. That's probably OK.

Tratcher commented:
- Rebased.
- Removed the TlsHandshakeCallbackOptions constructor.
- ServerOptionsSelectionCallback no longer sets AllowDelayedClientCertificateNegotation = true.

halter73 commented on src/Servers/Kestrel/Core/src/TlsHandshakeCallbackContext.cs:
```diff
@@ -0,0 +1,51 @@
+        /// <summary>
+        /// Information about an individual connection.
+        /// </summary>
+        public ConnectionContext Connection { get; internal set; } = default!;
```
👍 This resolves part of #31303 too.

The next major thing is to not break ALPN for HTTP/2 when this new UseHttps() overload is used. Exposing the ConfigureAlpn() is better than nothing, but I would prefer if it just works without needing to know to call that.

Tratcher committed:
Fix test
